app/models/sale.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class Sale < ApplicationRecord
+  belongs_to :seller, class_name: 'User', optional: true
+  belongs_to :client, class_name: 'User'
+  belongs_to :payment_method
+  belongs_to :invoice
+  has_one :subscription, dependent: :destroy
+  has_many :refunds, dependent: :destroy
+  has_many :articles_sales, dependent: :destroy
+  has_many :articles, through: :articles_sales
+  has_many :sales_subscription_offers, dependent: :destroy
+  has_many :subscription_offers, through: :sales_subscription_offers
+
+  accepts_nested_attributes_for :articles_sales
+
+  validates :total_price, presence: true
+
+  def verify
+    self.verified_at = Time.zone.now if verified_at.nil?
+  end
+
+  def generate(duration:, seller:)
+    return false if duration.to_i.negative?
+
+    return false unless generate_sales_subscription_offers duration.to_i
+
+    self.seller = seller
+    create_associated_subscription duration.to_i if duration.to_i.positive?
+    self.total_price = compute_total_price
+    verify if payment_method&.auto_verify
+    generate_invoice
+    true
+  end
+
+  def compute_total_price
+    total = 0
+    articles_sales.each do |rec|
+      total += rec.quantity * Article.find(rec.article_id).price
+    end
+    sales_subscription_offers.each do |rec|
+      total += rec.quantity * SubscriptionOffer.find(rec.subscription_offer.id).price
+    end
+    total
+  end
```

Comment:
We don't care about that at our scale because we are talking about one or two articles, but I think here we would make a request to the database per article/subscription. That can be a problem once you get to dozens of articles.
We don't need to change anything for our case though, it's simpler like that